filters/builtin/host.go

```diff
@@ -0,0 +1,128 @@
+}
+
+// Returns a new modhost filter Spec, whose instances execute
+// regexp.ReplaceAllString on the request host. Instances expect two
+// parameters: the expression to match and the replacement string.
+// Name: "modhost".
+func NewModHost() filters.Spec { return &modHost{behavior: regexpReplaceHost} }
+
+// Returns a new setHost filter Spec, whose instances replace
+// the request host.
+//
+// As an EXPERIMENTAL feature: the setHost filter provides the possiblity
+// to apply template operations. The current solution supports templates
+// with placeholders of the format: ${param1}, and the placeholders will
+// be replaced with the values of the same name from the wildcards in the
+// Path() predicate.
+//
+// See: https://godoc.org/github.com/zalando/skipper/routing#hdr-Wildcards
+//
+// The templating feature will stay in Skipper, but the syntax of the
+// templating may change.
+//
+// See also: https://github.com/zalando/skipper/issues/182
+//
+// Instances expect one parameter: the new host to be set, or the host
+// template to be evaluated.
+//
+// Name: "setHost".
+func NewSetHost() filters.Spec { return &modHost{behavior: fullReplaceHost} }
+
+// "modHost" or "setHost"
+func (spec *modHost) Name() string {
+	switch spec.behavior {
+	case regexpReplaceHost:
+		return ModHostName
+	case fullReplaceHost:
+		return SetHostName
+	default:
+		panic("unspecified behavior")
+	}
+}
+
+func createModHost(config []interface{}) (filters.Filter, error) {
+	if len(config) != 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	expr, ok := config[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	replacement, ok := config[1].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	rx, err := regexp.Compile(expr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &modHost{behavior: regexpReplaceHost, rx: rx, replacement: replacement}, nil
+}
+
+func createSetHost(config []interface{}) (filters.Filter, error) {
+	if len(config) != 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	tpl, ok := config[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	return &modHost{behavior: fullReplaceHost, template: eskip.NewTemplate(tpl)}, nil
+}
+
+// Creates instances of the modHost filter.
+//
+//lint:ignore ST1016 "spec" makes sense here and we reuse the type for the filter
+func (spec *modHost) CreateFilter(config []interface{}) (filters.Filter, error) {
+	switch spec.behavior {
+	case regexpReplaceHost:
+		return createModHost(config)
+	case fullReplaceHost:
+		return createSetHost(config)
+	default:
+		panic("unspecified behavior")
+	}
+}
+
+// Modifies the host with regexp.ReplaceAllString.
+func (f *modHost) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+	switch f.behavior {
+	case regexpReplaceHost:
+		req.Host = f.rx.ReplaceAllString(req.Host, f.replacement)
+	case fullReplaceHost:
+		req.Host = f.template.Apply(ctx.PathParam)
```

Comment:
What to do with this? Should we remove this whole `fullReplaceHost` functionality? Or can we somehow get `ctx.HostParam`?

Reply:
I guess it should be dropped, in case you decide to go with my `modeRequestHeader` idea.